acnportal/acnsim/base.py

```diff
@@ -0,0 +1,649 @@
+                                 of this object]}
+                    ```
+            Dict[str, JSON Serializable]: For clarity, the context_dict
+                is returned since it is modified by this function.
+
+        Warns:
+            UserWarning: If any attributes are present in the object
+                not handled by the object's `_to_dict` method.
+            UserWarning: If any of the unhandled attributes are not
+                JSON serializable.
+
+        """
+        first_call = context_dict is None
+        context_dict, = self._none_to_empty_dict(context_dict)
+        obj_id = f'{id(self)}'
+
+        # Check if this object has already been converted, and return
+        # the appropriate dict if this is the case.
+        if obj_id in context_dict:
+            return {'id': obj_id, 'context_dict': context_dict}, context_dict
+
+        obj_type = f'{self.__module__}.{self.__class__.__name__}'
+
+        # Get a dictionary of the attributes of this object using the
+        # object's _to_dict method.
+        attribute_dict, context_dict = self._to_dict(context_dict)
+
+        # Check that all attributes have been serialized.
+        # Warn if some attributes weren't serialized.
+        if attribute_dict.keys() != self.__dict__.keys():
+            unserialized_keys = \
+                set(self.__dict__.keys()) - set(attribute_dict.keys())
+            warnings.warn(
+                f"Attributes {unserialized_keys} present in object of "
+                f"type {obj_type} but not handled by object's _to_dict "
+                f"method. Serialized object may not load correctly. "
+                f"Write a _to_dict method and re-dump, or write an "
+                f"appropriate _from_dict method to accurately load.",
+                UserWarning
+            )
+            for key in unserialized_keys:
+                unserialized_attr = self.__dict__[key]
+                # Try calling the attr's _to_registry method.
+                try:
+                    # noinspection PyProtectedMember
+                    registry, context_dict = unserialized_attr._to_registry(
+                        context_dict=context_dict)
+                    attribute_dict[key] = registry['id']
+                    continue
+                except AttributeError:
+                    pass
+                # Try dumping the object using the native JSON
+                # serializer.
+                try:
+                    json.dumps(unserialized_attr)
+                except TypeError:
+                    warnings.warn(
+                        f"Attribute {key} could not be serialized. "
+                        f"Dumping the attribute's repr() "
+                        f"representation. This attribute will not be "
+                        f"fully loaded.",
+                        UserWarning
+                    )
+                    attribute_dict[key] = [__NOT_SERIALIZED_FLAG__,
```

Comment:
For objects that weren't serialized, I added a flag in the attribute_dict to distinguish such objects from e.g. strings, so that the deserializer knows when to use the ErrorAllWrapper.